File: protocol-generator/csharp/src/EntityCore.Protocol/AuthenticationException.cs
namespace EntityCore.Protocol;

/// <summary>
/// Authentication failed during connection establishment — an invalid nonce
/// signature, a public key that does not match the claimed peer id, or an
/// unsupported key type (V7 §4.6, §4.7). Maps to status 401 by default.
/// </summary>
public sealed class AuthenticationException : EntityProtocolException
{
    /// <summary>Create an authentication failure with its connection error code.</summary>
    public AuthenticationException(string code, string message, int status = 401)
        : base(message, status)
    {
        Code = code;
    }

    /// <summary>The §4.7 connection error code (e.g. <c>"invalid_signature"</c>).</summary>
    public string Code { get; }
}

File: protocol-generator/java/src/main/java/org/entitycore/protocol/peer/AuthenticationException.java
package org.entitycore.protocol.peer;

/**
 * §4.6 / §5.2 authentication failure — maps to wire status 401 at the dispatcher
 * boundary. The §5.5 unresolvable-grantee carve-out (a grantee that cannot be
 * resolved is a 401, not a 403) is also signalled with this type.
 */
public final class AuthenticationException extends EntityProtocolException {
    private static final long serialVersionUID = 1L;

    public AuthenticationException(String message) {
        super(message);
    }
}
